atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraRequestExceptionHandler.java

```diff
@@ -0,0 +1,169 @@
+/*
+ * Copyright 2017 Palantir Technologies, Inc. All rights reserved.
+ *
+ * Licensed under the BSD-3 License (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://opensource.org/licenses/BSD-3-Clause
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.keyvalue.cassandra;
+
+import java.net.InetSocketAddress;
+import java.net.SocketException;
+import java.net.SocketTimeoutException;
+import java.util.NoSuchElementException;
+import java.util.concurrent.ThreadLocalRandom;
+import java.util.function.Supplier;
+
+import org.apache.cassandra.thrift.InvalidRequestException;
+import org.apache.cassandra.thrift.TimedOutException;
+import org.apache.cassandra.thrift.UnavailableException;
+import org.apache.thrift.transport.TTransportException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.slf4j.helpers.MessageFormatter;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.palantir.atlasdb.keyvalue.api.InsufficientConsistencyException;
+import com.palantir.atlasdb.keyvalue.cassandra.CassandraClientFactory.ClientCreationFailedException;
+import com.palantir.logsafe.SafeArg;
+import com.palantir.logsafe.UnsafeArg;
+
+class CassandraRequestExceptionHandler {
+    private static final Logger log = LoggerFactory.getLogger(CassandraClientPool.class);
+
+    private final Supplier<Integer> maxTriesSameHost;
+    private final Supplier<Integer> maxTriesTotal;
+    private final Blacklist blacklist;
+
+    CassandraRequestExceptionHandler(
+            Supplier<Integer> maxTriesSameHost,
+            Supplier<Integer> maxTriesTotal,
+            Blacklist blacklist) {
+        this.maxTriesSameHost = maxTriesSameHost;
+        this.maxTriesTotal = maxTriesTotal;
+        this.blacklist = blacklist;
+    }
+
+    <K extends Exception> void handleRequest(
+                RetryableCassandraRequest<?, K> req,
+                InetSocketAddress hostTried,
+                Exception ex)
+            throws K {
+        req.incrementNumberOfAttempts();
+        req.triedOnHost(hostTried);
```

Comment:
These two calls seem heavily coupled - I think `triedOnHost` can reasonably `incrementNumberOfAttempts` itself, and then we only need one method call here.

Reply:
Done.